### PR TITLE
[REVIEW ME PLS] Battle Monsters Card Game, or the last* PR.

### PR DESCRIPTION
**Pull request by BurgerLUA**
# Overview
Adds Battle Monster cards. Battle Monsters is a collectable card game clone like Yugioh or magic the gathering. Players can purchase card packs and use them in games with actual rules to them.

## Checklist
Card Generation Framework: 100%
Sprites: 80%
Game Rules: 90%
Deck Framework: 5%
Monster Cards: 20%
Spell/Trap Cards: 5%
Obtainable Status: 0%

## WIP
Code feedback on things such as generation/implementation is welcome. I will not be accepting non-code feedback such as feature suggestions or requests.

### CODE REVIEW

**Comment by BotBOREALIS** [comment]
Icons with diff:
<details><summary>icons/obj/battle_monsters/card.dmi</summary>

Key | Old | New | Status
--- | --- | --- | ---
stack||stack|Created
stack[1]||![stack[1]](http://images.aurorastation.org/images/04de01bf019ae76e7739fef94e9ad450.png)|Created
stack[2]||![stack[2]](http://images.aurorastation.org/images/d3f8ffa462413bc40ea20c5327ef7e36.png)|Created
stack[3]||![stack[3]](http://images.aurorastation.org/images/d3f8ffa462413bc40ea20c5327ef7e36.png)|Created
back||back|Created
rarity_animation||rarity_animation|Created
rarity_animation[1]||![rarity_animation[1]](http://images.aurorastation.org/images/755195f7a74d3ff7e5f3f3b174fa9d63.png)|Created
rarity_animation[2]||![rarity_animation[2]](http://images.aurorastation.org/images/db17679ab6bd8cd730014609f6f68422.png)|Created
rarity_animation[3]||![rarity_animation[3]](http://images.aurorastation.org/images/69683c86e86c8a1576795b8dadd3a014.png)|Created
rarity_animation[4]||![rarity_animation[4]](http://images.aurorastation.org/images/f47b7d41eeb6650dbb5c3e32f7b2066c.png)|Created
rarity_animation[5]||![rarity_animation[5]](http://images.aurorastation.org/images/afcaa3b0e4c0b6b8e24f6c9efd0585db.png)|Created
rarity_animation[6]||![rarity_animation[6]](http://images.aurorastation.org/images/b14fe164430378b7baa517eef68bef14.png)|Created
rarity_animation[7]||![rarity_animation[7]](http://images.aurorastation.org/images/39d225fc13aacbca7b6c2063dcac2f5b.png)|Created
rarity_animation[8]||![rarity_animation[8]](http://images.aurorastation.org/images/6ad863956da16e6a301bc3ad01f4744d.png)|Created
rarity_animation[9]||![rarity_animation[9]](http://images.aurorastation.org/images/0703d27074ea6edd6d4d19828b1b55c8.png)|Created
rarity_animation[10]||![rarity_animation[10]](http://images.aurorastation.org/images/b55e19f7ab7c015e16fc136fd094347c.png)|Created
front_r_1||front_r_1|Created
front_r_2||front_r_2|Created
front_r_3||front_r_3|Created
front_r_4||front_r_4|Created
front_label||front_label|Created
|||Created
sword||sword|Created
wall||wall|Created
magic||magic|Created
narsie||narsie|Created
clown||clown|Created
banana||banana|Created
pie||pie|Created
potion_red||potion_red|Created
potion_blue||potion_blue|Created
ian_right||ian_right|Created
ian_left||ian_left|Created
ian_body||ian_body|Created
ian_face||ian_face|Created
stop||stop|Created
fire||fire|Created
energy||energy|Created
water||water|Created
ice||ice|Created
earth||earth|Created
stone||stone|Created
dark||dark|Created
light||light|Created
light_god||light_god|Created
dark_god||dark_god|Created
neutral_god||neutral_god|Created
true_god||true_god|Created
holy_beam||holy_beam|Created
swap||swap|Created
color||color|Created
boost||boost|Created
candle||candle|Created
burned||burned|Created
burning_sun||burning_sun|Created
scorching_blade||scorching_blade|Created
burning_sun_god||burning_sun_god|Created
powered||powered|Created
dragonslayer||dragonslayer|Created
dragoncult||dragoncult|Created
no_element||no_element|Created
staff||staff|Created
human||human|Created
human_knight||human_knight|Created
</details>

**Comment by ParadoxSpace** [comment]
PR author has failed to explain the * addendum.

**Comment by BurgerLUA** [comment]
This is the only remaining PR where considerable work has been put into it and it would be a waste for me to discard the project completely. It's also an addition that I look forward to and will possibly be the only thing that keeps me interested in playing on the server.

**Comment by BurgerLUA** [comment]
Don't worry about undocumented changes, everything will be documented when it's done and ready for an actual review.

File: code/modules/battlemonsters/datum_elements.dm [hunk 1]
```diff
@@ -0,0 +1,346 @@
...
+datum/battle_monsters/element/none
+	name = ""
+	id = "no_element"
+	icon_state = "no_element"
+	elements = BATTLE_MONSTERS_ELEMENT_NEUTRAL
+	description = ""
+	rarity = 5
+	rarity_score = -1
+
+datum/battle_monsters/element/trap //Indicator for trap cards.
```
Comment (HeyBanditoz) [inline code/modules/battlemonsters/datum_elements.dm]: Please for the love of all that is holy add a `/` before datum.

Comment (BurgerLUA) [reply code/modules/battlemonsters/datum_elements.dm]: is there a code related reason or is it cosmetic related

**Comment by BurgerLUA** [comment]
!review

**Comment by BurgerLUA** [comment]
It's actually done and ready for review.

File: code/game/objects/items/weapons/manuals.dm [hunk 2]
```diff
@@ -1416,4 +1416,33 @@
 					<p>201: Means that the frame is not rated to handle the entirety of the heat energy of the components. Either upgrade the frame, or downgrade the components.</p>
 				</body>
 			</html>
+			"}
+
+/obj/item/weapon/book/manual/battlemonsters
+	name = "/improper Guide to Battlemonsters"
```
Comment (Lohikar) [inline code/game/objects/items/weapons/manuals.dm]: 🤔

File: code/modules/battlemonsters/items/card.dm [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+	var/datum/battle_monsters/title/suffix_datum
+
+	var/datum/battle_monsters/spell/spell_datum
+	var/datum/battle_monsters/trap/trap_datum
+
+	w_class = ITEMSIZE_TINY
+
+	//Card information here
+
+/obj/item/battle_monsters/card/New(var/turf/loc,var/prefix,var/root,var/title,var/trap,var/spell)
```
Comment (Lohikar) [inline code/modules/battlemonsters/items/card.dm]: 👀

File: code/modules/battlemonsters/items/core.dm [hunk 4]
```diff
@@ -0,0 +1,106 @@
+/obj/item/battle_monsters/
+	icon = 'icons/obj/battle_monsters/card.dmi'
+	icon_state = ""
+	var/list/center_of_mass = list("x"=16, "y"=16)
+	var/facedown = TRUE
+	var/rotated = FALSE
+
+/obj/item/battle_monsters/dropped(mob/user as mob)
+	dir = user.dir
```
Comment (Lohikar) [inline code/modules/battlemonsters/items/core.dm]: Prob. should be set_dir() if you want these to play nice with openspace.

File: code/modules/battlemonsters/items/core.dm [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+	if(rotated)
+		dir = turn(dir,90)
+	update_icon()
+	. = ..()
+
+/obj/item/battle_monsters/pickup(mob/user as mob)
+	dir = NORTH
```
Comment (Lohikar) [inline code/modules/battlemonsters/items/core.dm]: Same here.

File: code/modules/battlemonsters/items/core.dm [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+	attack_self(user)
+
+/obj/item/battle_monsters/proc/RotateCard(var/mob/user)
+
+	rotated = !rotated
+
+	if(rotated)
+		if(src.loc == user)
+			to_chat(user,span("notice", "You prepare \the [name] to be played horizontally."))
+			dir = turn(NORTH,90)
```
Comment (Lohikar) [inline code/modules/battlemonsters/items/core.dm]: And here.

File: code/modules/battlemonsters/items/deck.dm [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+			span("notice","\The [user] combines two decks together."),\
+			span("notice","You combine too decks together.")\
+		)
+
+		qdel(src)
+		return
+
+	. = ..()
+
+/obj/item/battle_monsters/deck/verb/toggle_mode()
```
Comment (Lohikar) [inline code/modules/battlemonsters/items/deck.dm]: Might be a good idea to throw a use_check() into this verb.

File: code/modules/battlemonsters/items/deck.dm [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+		)
+		icon_state = "stack"
+	else
+		usr.visible_message(\
+			span("notice","\The [usr] turns their stack of cards into a hand."),\
+			span("notice","You turn your stack of cards into a hand.")\
+		)
+		icon_state = "hand"
+
+/obj/item/battle_monsters/deck/verb/shuffle()
```
Comment (Lohikar) [inline code/modules/battlemonsters/items/deck.dm]: And this one.

File: code/modules/battlemonsters/items/deck.dm [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+
+/obj/item/battle_monsters/deck/attackby(var/obj/item/attacking, var/mob/user)
+	if(istype(attacking,/obj/item/battle_monsters/card) && attacking != src)
+		var/obj/item/battle_monsters/card/adding_card = attacking
+		add_card(user,adding_card)
+
+/obj/item/battle_monsters/deck/attack_self(mob/user)
+
+	if(user != src.loc) //Idk how this is possible but you never know.
+		to_chat(user,span("notice","You'll have to pick up \the [src] to examine the cards!"))
```
Comment (Lohikar) [inline code/modules/battlemonsters/items/deck.dm]: You show an error message, but don't actually do anything about it.

File: code/modules/battlemonsters/items/furniture/dueling_area.dm [hunk 10]
```diff
@@ -0,0 +1,49 @@
+/obj/structure/dueling_table
+	name = "dueling table"
+	icon = 'icons/obj/battle_monsters/furniture.dmi'
+	anchored = 1
+	density = 1
+	climbable = 1
+	throwpass = 1
+
+/obj/structure/dueling_table/no_colide
```
Comment (Lohikar) [inline code/modules/battlemonsters/items/furniture/dueling_area.dm]: Collide.

File: code/modules/battlemonsters/items/wrapped.dm [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+	name = "battle monsters starterpack"
+	desc = "A battle monsters 52 card deck starterpack. Contains 52 basic cards and a lifetime of sadness."
+	icon_state = "pack1"
+	var/contained_cards = 52
+	var/rare_limiter = 1000
+
+/obj/item/battle_monsters/wrapped/attack_self(mob/user)
+	var/list/deck_data = list()
+	for(var/i=1,i < contained_cards,i++)
+		CHECK_TICK //This stuff is a little intensive I think.
```
Comment (Lohikar) [inline code/modules/battlemonsters/items/wrapped.dm]: 👀

File: code/game/objects/items/weapons/manuals.dm (hunk 2 above)
Comment (BurgerLUA) [reply code/game/objects/items/weapons/manuals.dm]: /improper signals improper known, right?

File: code/modules/battlemonsters/items/wrapped.dm (hunk 11 above)
Comment (Lohikar) [reply code/modules/battlemonsters/items/wrapped.dm]: No, it's just that I feel like something like this shouldn't be expensive enough that a C_T is required. Does the generation code do a lot of string processing or something?

Comment (BurgerLUA) [reply code/modules/battlemonsters/items/wrapped.dm]: I'm not entirely sure. It just gets a random element from a list using pickweight 3 times, 52 times.

File: code/modules/battlemonsters/items/deck.dm (hunk 7 above)
Comment (BurgerLUA) [reply code/modules/battlemonsters/items/deck.dm]: Done

**Comment by BurgerLUA** [comment]
There are a few card types that need to be added. Each element is supposed to have it's own god card, and currently only the fire element has its own god card. Once the reviews are in I'll start coding them.

Also the rules need to be implemented. The rules are being modified and improved currently. They need to be formatted into HTML first.

File: code/modules/battlemonsters/datum_elements.dm (hunk 1 above)
Comment (HeyBanditoz) [reply code/modules/battlemonsters/datum_elements.dm]: Searching and cosmetic

File: code/modules/battlemonsters/datum_elements.dm [hunk 12]
```diff
@@ -110,7 +113,7 @@
 	icon_state = "thunder"
 	power_add = 500
 	attack_add = 0.25
-	description = "A symbol on their forehead indicates that they are a %SPECIES thunderchild, born of the incredibly horny god Bleus after coitus with a %SPECIES."
+	description = "A symbol on their forehead indicates that they are a %SPECIES thunderchild, born of the incredibly horny god Bleus after coitus with a %SPECIES. Children born of Bleus tend to show a god-like control over energy elements, such as lightning."
```
Comment (Lohikar) [inline code/modules/battlemonsters/datum_elements.dm]: D E E P  L O R E

File: code/modules/battlemonsters/datum_elements.dm [hunk 13]
```diff
@@ -146,41 +149,63 @@
 	rarity = BATTLE_MONSTERS_RARITY_RARE
 	rarity_score = 1
 
+/datum/battle_monsters/element/energy/dragongod
+	name = "Dragon God"
+	id = "dragon_god"
+	icon_state = "dragon_god"
+	defense_type = BATTLE_MONSTERS_DEFENSETYPE_GOD | BATTLE_MONSTERS_DEFENSETYPE_GIANT_DRAGON
+	elements = BATTLE_MONSTERS_ELEMENT_ENERGY
+	description = "Their %ELEMENT_AND wings are absolutely massive, and their breath constantly exhales deadly %ELEMENT_AND. This being is a common god among dragons, and it is theorized that one of the many breeds of dragons came from %NAME."
```
Comment (Lohikar) [inline code/modules/battlemonsters/datum_elements.dm]: Their breath exhales?

File: code/modules/battlemonsters/datum_elements.dm [hunk 14]
```diff
@@ -146,41 +149,63 @@
...
 	power_mul = 0.75
 	attack_add = 0.5
 	rarity = BATTLE_MONSTERS_RARITY_UNCOMMON
 
 /datum/battle_monsters/element/water/trench_pirate
-	name = "Trench Pirate"
+	name = "Trench"
 	id = "trench_pirate"
-	description = "They're covered entirely with seaweed and barnacles."
+	description = "They're covered entirely with seaweed and barnacles. They wear an eyepatch and a bandana, and wield a %WEAPON_AND fearsomely. Rumor has it that Trench beings were shipwrecked sailors whose bodies sunk into marine trenches into the hells below."
```
Comment (Lohikar) [inline code/modules/battlemonsters/datum_elements.dm]: Latter sentence here is a bit awkward. Maybe find a way to phrase it without two intos?

File: code/modules/battlemonsters/datum_elements.dm [hunk 15]
```diff
@@ -194,22 +219,56 @@
...
 	attack_type = BATTLE_MONSTERS_ATTACKTYPE_CLAWS
 	power_add = 200
 	rarity = BATTLE_MONSTERS_RARITY_RARE
 	rarity_score = 1
 
+/datum/battle_monsters/element/ice/frostlord
+	name = "Frostlord"
+	id = "frostlord"
+	icon_state = "ice"
+	description = "They're wearing a large crown made out of ice and diamonds, and a royal cape made out magical yeti fur. Frostlords gain their title by defeating a yeti in combat by wrestling it to death."
```
Comment (Lohikar) [inline code/modules/battlemonsters/datum_elements.dm]: Missing word after `out`. Latter sentence seems a little awkward too, but I can't think of a way to fix it off the top of my head.

File: code/modules/battlemonsters/datum_elements.dm [hunk 16]
```diff
@@ -194,22 +219,56 @@
...
+	special_effects = "Royal Decree: As long as there are 2 other frost element monsters on the same side of the field, %NAME cannot be sent to the graveyard.<br>Spell Immunity: %NAME cannot be affected by traps or spells."
+	elements = BATTLE_MONSTERS_ELEMENT_ICE
+	power_add = 1000
+	rarity = BATTLE_MONSTERS_RARITY_RARE
+
+/datum/battle_monsters/element/ice/frost_guardian
+	name = "Frost Guardian"
+	id = "frostguardian"
+	icon_state = "ice"
+	description = "They're wearing incredibly bulky armor, made entirely out of ice. Their still posture compliments the age-long growing of the iciles forming on their %WEAPON_AND."
```
Comment (Lohikar) [inline code/modules/battlemonsters/datum_elements.dm]: `iciles`

File: code/modules/battlemonsters/datum_elements.dm [hunk 17]
```diff
@@ -194,22 +219,56 @@
...
+	special_effects = "Best Defense: As long as %NAME is on the field, its owner cannot lose life points from monster battles."
+	attack_mul = 0
+	elements = BATTLE_MONSTERS_ELEMENT_ICE
+	power_add = 1000
+	rarity = BATTLE_MONSTERS_RARITY_RARE
+
+/datum/battle_monsters/element/ice/blizard_god
+	name = "Blizard God"
```
Comment (Lohikar) [inline code/modules/battlemonsters/datum_elements.dm]: Shouldn't this be blizzard?

File: code/modules/battlemonsters/datum_elements.dm [hunk 18]
```diff
@@ -251,6 +310,19 @@
 	rarity = BATTLE_MONSTERS_RARITY_RARE
 	rarity_score = 1
 
+/datum/battle_monsters/element/earth/forest_god
+	name = "Forest God"
+	id = "frost_god"
```
Comment (Lohikar) [inline code/modules/battlemonsters/datum_elements.dm]: forest, perhaps?

File: code/modules/battlemonsters/datum_elements.dm [hunk 19]
```diff
@@ -251,6 +310,19 @@
...
+	icon_state = "earth_god"
+	elements = BATTLE_MONSTERS_ELEMENT_EARTH | BATTLE_MONSTERS_ELEMENT_WATER | BATTLE_MONSTERS_ELEMENT_STONE
+	defense_type = BATTLE_MONSTERS_DEFENSETYPE_GOD | BATTLE_MONSTERS_DEFENSETYPE_GIANT
+	description = "They have roots for feet, and massive tree trunts for arms. Their overall size rivals those of the greatest manmade structures on earth. They carry around a giant %WEAPON_AND made sculpted entirely out of the largest trees of the Great Forest."
```
Comment (Lohikar) [inline code/modules/battlemonsters/datum_elements.dm]: `trunts`; the last bit is awkward, consider dropping `made`.

File: code/modules/holodeck/HolodeckControl.dm [hunk 20]
```diff
@@ -281,6 +281,11 @@
 	for(var/obj/holo_obj in holographic_objs)
 		holo_obj.alpha *= 0.8 //give holodeck objs a slight transparency
 
+	if(HP.loop_ambience)
```
Comment (Lohikar) [inline code/modules/holodeck/HolodeckControl.dm]: Why not just set `loop_ambience` to this unconditionally? All the if is doing is preventing it from being null, which if you _really_ need you can do with an or.

**Comment by BurgerLUA** [comment]
This is pretty much done. However with all RNG stuff, there may be some grammatical issues and other things that will need to be resolved through an actual playtest. Thankfully none of the mechanics here have any impact on actual gameplay so it's not like people will complain.

File: code/modules/battlemonsters/items/card.dm [hunk 21]
```diff
@@ -0,0 +1,175 @@
...
```
Comment (skull132) [inline code/modules/battlemonsters/items/card.dm]: Why does this override even exist?

Comment (BurgerLUA) [reply code/modules/battlemonsters/items/card.dm]: I thought it solved a bug because initialize wasn't working without it.

**Comment by BurgerLUA** [comment]
Works.

**Pull request merged**